classes/services/PKPSubmissionFileService.inc.php

```diff
@@ -0,0 +1,837 @@
+				if ($uploader->getId() != $submissionFile->getData('uploaderUserId')) {
+					$uploader = Services::get('user')->get($submissionFile->getData('uploaderUserId'));
+				}
+
+				// Fetch the latest notification email timestamp
+				import('lib.pkp.classes.log.SubmissionEmailLogEntry'); // Import email event constants
+				$submissionEmailLogDao = DAORegistry::getDAO('SubmissionEmailLogDAO'); /* @var $submissionEmailLogDao SubmissionEmailLogDAO */
+				$submissionEmails = $submissionEmailLogDao->getByEventType($submission->getId(), SUBMISSION_EMAIL_AUTHOR_NOTIFY_REVISED_VERSION);
+				$lastNotification = null;
+				$sentDates = [];
+				if ($submissionEmails){
+					while ($email = $submissionEmails->next()) {
+						if ($email->getDateSent()){
+							$sentDates[] = $email->getDateSent();
+						}
+					}
+					if (!empty($sentDates)){
+						$lastNotification = max(array_map('strtotime', $sentDates));
+					}
+				}
+
+				import('lib.pkp.classes.mail.SubmissionMailTemplate');
+				$mail = new \SubmissionMailTemplate($submission, 'REVISED_VERSION_NOTIFY');
+				$mail->setEventType(SUBMISSION_EMAIL_AUTHOR_NOTIFY_REVISED_VERSION);
+				$mail->setReplyTo($context->getData('contactEmail'), $context->getData('contactName'));
+				// Get editors assigned to the submission, consider also the recommendOnly editors
+				$userDao = DAORegistry::getDAO('UserDAO'); /* @var $userDao UserDAO */
+				$editorsStageAssignments = $stageAssignmentDao->getEditorsAssignedToStage($submission->getId(), $reviewRound->getStageId());
+				foreach ($editorsStageAssignments as $editorsStageAssignment) {
+					$editor = $userDao->getById($editorsStageAssignment->getUserId());
+					 // IF no prior notification exists
+					 // OR if editor has logged in after the last revision upload
+					 // OR the last upload and notification was sent more than a day ago,
+					 // THEN send a new notification
+					if (is_null($lastNotification) || strtotime($editor->getDateLastLogin()) > $lastNotification || strtotime('-1 day') > $lastNotification){
+						$mail->addRecipient($editor->getEmail(), $editor->getFullName());
+					}
+				}
+				// Get uploader name
+				$mail->assignParams(array(
+					'authorName' => $uploader->getFullName(),
+					'editorialContactSignature' => $context->getData('contactName'),
+					'submissionUrl' => $request->getDispatcher()->url(
+						$request,
+						ROUTE_PAGE,
+						null,
+						'workflow',
+						'index',
+						[
+							$submission->getId(),
+							$reviewRound->getStageId(),
+						]
+					),
+				));
+
+				if ($mail->getRecipients()){
+					if (!$mail->send($request)) {
+						import('classes.notification.NotificationManager');
+						$notificationMgr = new \NotificationManager();
+						$notificationMgr->createTrivialNotification($request->getUser()->getId(), NOTIFICATION_TYPE_ERROR, ['contents' => __('email.compose.error')]);
+					}
+				}
+			}
+		}
+
+		HookRegistry::call('SubmissionFile::add', [$submissionFile, $request]);
```

Comment:
Not critical, but I think `$submissionFile` should be passed by reference here. That would allow it to be reassigned wholesale by a hook registrant, not just modified.

Reply:
Hmm, ok. Maybe that's my misunderstanding. So if a hook did this:

```
HookRegistry::register('SubmissionFile::add', function($hookName, $args) {
  $submissionFile = $args[0];
  $newSubmissionFile = Services::get('submissionFile')->get(...);
  $submissionFile = $newSubissionFile;
}
```

In this case, would the `$submissionFile` not be changed in `PKPSubmissionFileService::add()`? If not, then you're right and this needs to be updated in all service classes.

Reply:
Yes, that's right.

Reply:
Ok good. Applied that to a _bunch_ of hooks. :)